signer/src/block_observer.rs

```diff
@@ -243,60 +247,104 @@ impl<C: Context, B> BlockObserver<C, B> {
         Ok(())
     }
 
-    /// Find the parent blocks from the given block that are also missing from our database
+    /// Set the sbtc start height, if it has not been set already.
+    async fn set_sbtc_start_height(&self) -> Result<(), Error> {
+        if self.context.state().is_sbtc_start_height_set() {
+            return Ok(());
+        }
+
+        let pox_info = self.context.get_stacks_client().get_pox_info().await?;
+        let nakamoto_start_height = pox_info
+            .nakamoto_start_height()
+            .ok_or(Error::MissingNakamotoStartHeight)?;
+
+        self.context
+            .state()
+            .set_sbtc_start_height(nakamoto_start_height);
+
+        Ok(())
+    }
+
+    /// Find the parent blocks from the given block that are also missing
+    /// from our database.
+    ///
+    /// # Notes
+    ///
+    /// This function does two things:
+    /// 1. Set the `sbtc_start_height` if it has not been set already. If
+    ///    it is not set, then we fetch the stacks nakamoto start height
+    ///    from stacks-core and use that value.
+    /// 2. Continually fetches block headers from bitcoin-core until it
+    ///    encounters a known block header or if the height of the block is
+    ///    less than or equal to the `sbtc_start_height`.
+    ///
+    /// If there are many unknown blocks then this function can take some
+    /// time. Since each header is 80 bytes, we should be able to fetch
+    /// headers for the entire bitcoin blockchain (~900k blocks at the time
+    /// of writing) into memory.
     #[tracing::instrument(skip_all, fields(%block_hash))]
-    async fn next_blocks_to_process(
+    pub async fn next_headers_to_process(
```

Comment:
I really think that this sync/process logic should be somewhere else and callable both at startup and on observed block.

When running the initial sync here, then the signer is online and will attempt to respond to potential signing requests which will unnecessarily fail (bad chain tip). It's not a problem per-se I don't think, but it just doesn't feel "clean" and mucks up the logs.

Maybe I'm just oldschool, but I like to ensure that things are in a good/known state before starting up -- Stacks core does this too, for example, making sure it's synced bitcoin block headers before it starts doing stuff.

Reply:
Just want to add that this isn't a blocker for me to merge the PR, but something I think we should discuss

Reply:
Makes sense, I'm not opposed. I just didn't want to make the change here. Most of the decisions here are about keeping the diff as small as possible before launch.

Reply:
Yeah we need to push this out of this PR for the same reasons mentioned in https://github.com/stacks-network/sbtc/pull/1093#discussion_r1880266567, which is that doing this start-up isn't so simple. Probably doable, just not super straightforward.